pkg/controller/service_instance.go

```
@@ -0,0 +1,183 @@
+	var instanceSpec sc_v1b1.ServiceInstance
+	var previousEncodedChecksum string
+	var updateCount int64
+
+	if err := unstructured_conversion.DefaultConverter.FromUnstructured(spec.Object, &instanceSpec); err != nil {
+		return nil, errors.Wrapf(err, "Failure to parse ServiceInstance spec %q", spec.GetName())
+	}
+
+	if len(instanceSpec.Spec.ParametersFrom) == 0 {
+		return spec, nil
+	}
+
+	if v, ok := instanceSpec.ObjectMeta.Annotations[secretParametersChecksumAnnotation]; ok && v == disabled {
+		return spec, nil
+	}
+
+	if actual != nil {
+		spec, ok := actual.(*sc_v1b1.ServiceInstance)
+		if !ok {
+			return nil, errors.New("Failure retrieving ServiceInstance spec")
+		}
+		previousEncodedChecksum = spec.ObjectMeta.Annotations[secretParametersChecksumAnnotation]
+		updateCount = spec.Spec.UpdateRequests
+	}
+
+	checkSum, err := st.calculateNewCheckSum(&instanceSpec, namespace, previousEncodedChecksum, updateCount)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to generate new checksum for spec %q", spec.GetName())
+	}
+
+	if previousEncodedChecksum != "" && checkSum != previousEncodedChecksum {
+		instanceSpec.Spec.UpdateRequests = updateCount + 1
+	}
+
+	instanceSpec = setAnnotation(instanceSpec, checkSum)
+
+	unstructuredSpec, err := util.RuntimeToUnstructured(&instanceSpec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to marshal back into unstructed %q", spec.GetName())
+	}
+
+	return unstructuredSpec, nil
+}
+
+func (st *resourceSyncTask) calculateNewCheckSum(instanceSpec *sc_v1b1.ServiceInstance, namespace string, previousEncodedChecksum string, updateCount int64) (string, error) {
+	checksumPayload, err := st.generateSecretParametersChecksumPayload(&instanceSpec.Spec, namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if previousEncodedChecksum != "" {
+		previousCheckSum, err := decodeChecksum(previousEncodedChecksum)
```

Invalid data in the annotation should not stop the processing - just ignore the error and overwrite the invalid value.

---

Maybe log the error